File: client/src/api/recommend.ts
import client from './client';
import type { RecommendResult } from '@/types/recommend';

export type RecommendRequest = {
  workplace_name: string;
  workplace_address: string;
  job_type: string;
  max_commute_minutes: number;
  lifestyle_tags: string[];
  deposit_max_wan: number;
  monthly_rent_max_wan: number;
  age: number;
  annual_income_wan: number;
  household_type: string;
  home_ownerless: boolean;
};

export async function postRecommend(body: RecommendRequest): Promise<RecommendResult> {
  console.log('[api] postRecommend body', body);
  const { data } = await client.post<RecommendResult>('/api/v1/recommend', body);
  console.log('[api] postRecommend response areas:', data.areas?.length, 'match_id:', data.match_id);
  return data;
}

// 세션 기반 최근 추천 복원 (새로고침/재진입 대비)
export async function getLatestRecommend(): Promise<RecommendResult> {
  const { data } = await client.get<RecommendResult>('/api/v1/recommend/latest');
  console.log('[api] getLatestRecommend areas:', data.areas?.length, 'match_id:', data.match_id);
  return data;
}

File: server/app/api/v1/recommend.py
import json

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from sqlalchemy import text

from app.core.auth import SessionDep
from app.core.deps import ConnDep
from app.core.exceptions import AddressNotFound
from app.schemas.area_recommend import RecommendResponse
from app.services import recommend as recommend_service

router = APIRouter(tags=["recommend"])


class RecommendRequest(BaseModel):
    workplace_name: str
    workplace_address: str
    job_type: str
    max_commute_minutes: int
    lifestyle_tags: list[str]
    deposit_max_wan: int
    monthly_rent_max_wan: int
    age: int
    annual_income_wan: int
    household_type: str
    home_ownerless: bool


@router.post(
    "/recommend",
    status_code=200,
    response_model=RecommendResponse,
    response_model_exclude_unset=True,
)
async def post_recommend(
    req: RecommendRequest,
    session: SessionDep,
    conn: ConnDep,
) -> RecommendResponse:
    try:
        result = await recommend_service.recommend(req, conn, session.session_id)
    except AddressNotFound:
        raise HTTPException(status_code=400, detail="address_not_found") from None
    if isinstance(result, list):
        return RecommendResponse(areas=result)
    return RecommendResponse(areas=result["areas"], match_id=result["match_id"])


@router.get("/recommend/latest", status_code=200, response_model=RecommendResponse)
async def get_recommend_latest(session: SessionDep, conn: ConnDep) -> RecommendResponse:
    stmt = text(
        "SELECT mr.id AS id, mr.payload AS payload, mr.created_at AS created_at"
        " FROM match_results mr"
        " JOIN profiles p ON p.id = mr.profile_id"
        " WHERE (p.session_id = :sid OR (p.user_id IS NOT NULL"
        " AND p.user_id = (SELECT user_id FROM sessions WHERE id = :sid)))"
        " ORDER BY mr.created_at DESC LIMIT 1"
    )
    _result = await conn.execute(stmt, {"sid": session.session_id})
    row = _result.mappings().first()
    if row is None:
        return RecommendResponse(areas=[], match_id=None)
    payload = row["payload"]
    data = payload if isinstance(payload, dict) else json.loads(payload)
    areas = data.get("areas", [])
    request_obj = data.get("request")
    return RecommendResponse(
        areas=areas, match_id=row["id"], created_at=row["created_at"], request=request_obj
    )
